### PR TITLE
Invalidate stale-while-revalidate cache entries on write

### PR DESCRIPTION
**Pull request by Will-Howard**
https://github.com/ForumMagnum/ForumMagnum/pull/9204 adds the ability to use `stale-while-revalidate` caching of post pages in a CDN. With the code in that PR, any updates to a post will not be reflected in the cached version until the page has been fetched once (so after a post is updated, one user would get a stale version, and the next user would get a version with the updates). This PR adds callbacks to refresh the cache entry on certain updates, so any major changes to the page will be shown to the next user to load the page (after ~1 minute).

I haven't tried to trigger invalidation on _every_ update, for instance votes on comments don't trigger invalidation. I used the places where we call `elasticSyncDocument` as a guide for when major updates are made. These are:
- Whenever a post is created or updated (via a graphql mutation, or `updateMutator` being called, but _not_ raw db updates)
- Whenever a post is voted on
- Added on top of the `elasticSyncDocument` cases: Whenever a comment is created or updated

Update: I was under a misapprehension about how CloudFront's caching works which means the previous version wouldn't work to clear the stale page. As it turns out, CloudFront (as set up for us) has 3 layers of caching, edge locations, regional edge caches, and an "origin shield" (of which there is only one, which enables global caching even in rarely used regions/edge locations):

Screenshot 2024-05-20 at 11 12 04

Pinging the url will trigger a refresh for the edge location that it hits, and the corresponding regional edge location, and the origin shield. But, if the page was also cached in another edge location, the first request to that location will still get the stale version (revalidation only goes up the hierarchy but not down).

To fix this I have updated it to use proper [cloudfront invalidations](https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/Invalidation.html#invalidating-objects-api), which do clear all the edge caches, and then ping the url once we are sure that is complete (to ensure that the next user request gets a cached version)

For testing purposes, you can try voting on a post, and you should see the invalidation be created [here](https://us-east-1.console.aws.amazon.com/cloudfront/v4/home?region=us-east-1#/distributions/E281TFAEMO151D/invalidations) (if using the credentials that are commented out in loadDevDbSettings).

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1201302964208280/1207285831217856) by [Unito](https://www.unito.io)

### CODE REVIEW

**Review by oetherington — APPROVED** [review]
LGTM

File: packages/lesswrong/server/cache/swr.ts [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+
+const invalidateUrlFromQueue = async (): Promise<void> => {
+  const logger = loggerConstructor(`swr-invalidation-queue`)
+  const url = invalidationQueue.shift()
+  if (!url) return;
+
+  logger(`Sending invalidation request. URL: ${url}, serverId: ${serverId}`)
+  fetch(url, {
+    headers: {
+      'User-Agent': INVALIDATION_USER_AGENT
```
[the remark this answers is not in the document]
Comment (Will-Howard) [reply packages/lesswrong/server/cache/swr.ts]: This user agent doesn't have any special ability to bypass the cache, any request would do the same thing. The user agent is just there so it's easy to see in the logs (if e.g. these requests are making up a large fraction of all requests)

File: packages/lesswrong/server/cache/swr.ts [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+    }
+  }).then(response => {
```
[the remark this answers is not in the document]
Comment (Will-Howard) [reply packages/lesswrong/server/cache/swr.ts]: No reason, I've changed it now

**Comment by Will-Howard** [comment]
Merging in order to test on staging, I expect there may be some kinks to work out

**Pull request merged**